process/structure.go

```diff
@@ -0,0 +1,37 @@
+package process
+
+import (
+	"encoding/json"
+	"io"
+	"strings"
+	"time"
+)
+
+// Structure outputs messages written to it in a structured format, including metadata.
+type Structure struct {
+	w        io.Writer
+	metadata map[string]string
+	encoder  *json.Encoder
+}
+
+func NewStructure(w io.Writer, metadata map[string]string) *Structure {
+	var metadata_copy = map[string]string{}
+	for k, v := range metadata {
+		metadata_copy[k] = v
+	}
+	return &Structure{w, metadata_copy, json.NewEncoder(w)}
+}
+
+// Write will write the given data out to the associated writer in a serialized JSON format, with fields including metadata and timestamp.
+func (s *Structure) Write(data []byte) (n int, err error) {
+	// When writing as a structured log, trailing newlines and carriage returns generally don't make sense.
+	var msg = strings.TrimRight(string(data), "\r\n")
+	var log = map[string]string{"message": msg, "timestamp": time.Now().UTC().Format(time.RFC3339)}
+	for k, v := range s.metadata {
+		log[k] = v
+	}
+	if err = s.encoder.Encode(log); err != nil {
+		return 0, err
+	}
```

Comment:
i was thinking of making a comment to the tune of "should we be calling `json.Marshal` (or equivalent) for every line? is the method used in the agent json logger more efficient?", but as it turns out, using the json library is notably more efficient:

```Go
package main_test

import (
	"encoding/json"
	"fmt"
	"strings"
	"testing"
)

func makeTestMap() map[string]string {
  // ...
}

func BenchmarkUnmarshall(b *testing.B) {
	for i := 0; i < b.N; i++ {
		mm := makeTestMap()
		marsh, _ = json.Marshal(mm)
	}
}

func BenchmarkStringBuilder(b *testing.B) {
	for i := 0; i < b.N; i++ {
		mm := makeTestMap()
		b := strings.Builder{}
		for k, v := range mm {
			b.WriteString(fmt.Sprintf("%q:%q,", k, v))
		}

		s = fmt.Sprintf("{%s}", strings.TrimSuffix(b.String(), ","))
	}
}
```

```
^_^ (~/b/scratch-go) go test -bench=.
goos: darwin
goarch: arm64
pkg: github.com/moskyb/scratch-go
BenchmarkUnmarshall-10       	  491529	      2445 ns/op
BenchmarkStringBuilder-10    	  365670	      3185 ns/op
PASS
ok  	github.com/moskyb/scratch-go	3.413s
```